raco/backends/myria/myria.py

```diff
@@ -1721,6 +1824,193 @@ def fire(self, op):
 
 
+class FillInJoinPullOrder(rules.Rule):
+
+    def __init__(self, pull_order):
+        self._disabled = False
+        self.pull_order = pull_order
+
+    def fire(self, op):
+        if not isinstance(op, MyriaSymmetricHashJoin):
+            return op
+        op.pull_order = self.pull_order
+        return op
+
+
+class PropagateAsyncFTBuffer(rules.Rule):
+
+    def __init__(self):
+        self._disabled = False
+
+    def fire(self, op):
+        if not isinstance(op, MyriaIDBController):
+            return op
+        buffer_type = op.get_agg_json()
+        if buffer_type['type'] == "CountFilter":
+            # do not propagate CountFilter
+            return op
+        # only do it for the iterative input
+        if isinstance(op.args[1], MyriaShuffleConsumer):
+            op.args[1].input.buffer_type = buffer_type
+        return op
+
+
+class ReplaceStoreIDB(rules.Rule):
+
+    @staticmethod
+    def collect_and_replace(op, idbproducers):
+        if not isinstance(op, algebra.NaryOperator):
+            return
+        if isinstance(op, algebra.UntilConvergence):
+            for idbproducer in op.children():
+                if isinstance(idbproducer.input, MyriaIDBController):
+                    idbproducers[idbproducer.input.name] = idbproducer
+            return
+        newchildren = []
+        for child in op.children():
+            if isinstance(child, algebra.UntilConvergence):
+                ReplaceStoreIDB.collect_and_replace(child, idbproducers)
+            if isinstance(child, algebra.Store) and \
+                    isinstance(child.input, algebra.ScanIDB):
+                assert child.input.name in idbproducers
+                idbproducers[child.input.name].input.relation_key = \
+                    child.relation_key
+            else:
+                newchildren.append(child)
+        op.args = newchildren
+
+    def fire(self, op):
+        self.collect_and_replace(op, {})
+        return op
+
+
+class RemoveEmptyFilter(rules.Rule):
+
+    def fire(self, expr):
+        for idx, op in enumerate(expr.children()):
+            if isinstance(op, algebra.Select) and op.condition is None:
+                child = op.children()[0]
+                if isinstance(expr, algebra.UnaryOperator):
+                    expr.input = child
+                elif isinstance(expr, algebra.BinaryOperator):
+                    if idx == 0:
+                        expr.left = child
+                    else:
+                        expr.right = child
+                elif isinstance(expr, algebra.NaryOperator):
+                    expr.args[idx] = child
+        return expr
+
+
+class RemoveSingleSplit(rules.Rule):
+
+    def fire(self, op):
+        if not isinstance(op, algebra.UntilConvergence):
+            return op
+
+        parent_map = {}
+        op.collectParents(parent_map)
+        for idx, child in enumerate(op.children()):
+            if not isinstance(child, MyriaSplitProducer):
```

Comment:
`UntilConvergence` has a list of `MyriaSplitProducer` sthat sit on top of `IDBController`s. It is actually also a child of `MyriaConsumer`, but for breaking the infinite optimization loop (see above), I made `MyriaConsumer` as a `ZeroaryOperator` with its producer as a field but as the child (`input`).